### ISSUE TITLE
Memory leak tables corrupted when calling DLL

### ISSUE THREAD

**Issue by uecasm** [cpputest/cpputest#1279]
I haven't worked out precisely why as yet, but given the following code (on Windows VS2017 with Boost 1.68 using DLLs for both the CRT and Boost, and in Debug config):

```
#include <sstream>
#include <boost/filesystem.hpp>

#include <CppUTest/TestHarness.h>
#include <CppUTestExt/MockSupport.h>
#include <CppUTest/CommandLineTestRunner.h>
#include <CppUTest/TestRegistry.h>
#include <CppUTestExt/MockSupportPlugin.h>
#include <CppUTestExt/MemoryReporterPlugin.h>
#include <CppUTest/MemoryLeakDetector.h>

#include <CppUTest/MemoryLeakDetectorNewMacros.h>
#include <CppUTest/MemoryLeakDetectorMallocMacros.h>

TEST_GROUP(TestTest)
{
};

TEST(TestTest, Test)
{
    for (auto i = 0; i < 100; ++i)
    {
        std::ostringstream str;
        str << "hello world " << i;

        boost::filesystem::path p(str.str());
        if (exists(p))
        {
            FAIL("why did that exist?");
        }
    }
}

int main(int argc, char *argv[])
{
    std::string x("yay");   // force a memory allocation prior to calling the below; see https://github.com/cpputest/cpputest/issues/1278

    MemoryLeakWarningPlugin::turnOnThreadSafeNewDeleteOverloads();

    MockSupportPlugin mockPlugin;
    MemoryReporterPlugin memoryPlugin;
    TestRegistry::getCurrentRegistry()->installPlugin(&mockPlugin);
    TestRegistry::getCurrentRegistry()->installPlugin(&memoryPlugin);
    return CommandLineTestRunner::RunAllTests(argc, argv);
}
```

Running this causes the `MemoryLeakDetectorTable` to get corrupted somehow and it ends up with the internal linked lists pointing into deleted memory, which will eventually cause a crash (often with some entirely unrelated code in the call stack, which made finding this difficult).

(Note that it tends to run successfully in Release builds, presumably because the Debug CRT poisons deleted memory.  It's also less noticeable when there are fewer allocations, which is why it's run in a loop.)

My best guess for the cause of this is some memory that is allocated on the app side via the CppUTest allocator, then passed to the DLL and freed on that side (not using the CppUTest allocator).  But I can't see why that would actually happen in this code; as far as I can tell the `path` is entirely on the app side (as it's mostly inlined); `exists` is in the Boost DLL, but it takes the parameter by const ref, so shouldn't be trying to reallocate it.

It does seem like the magic bit of code that breaks it is related to `path`s converting constructor (which calls `path_traits::dispatch` -- but that's a template so again it should be inlined).

I don't know if there's much that can actually be "fixed" in CppUTest for this case.

One thing that might help though: currently it looks like the allocation tracking data is appended to the memory block that's actually returned to the app.  This means that something outside of CppUTest can successfully delete that pointer, since it's still the start of a known allocated block.  It might be better to allocate the tracking data separately, or else to prepend the data to the user block (being careful of type alignment) instead.  That way if something tries to use the wrong allocator to delete the memory, it won't be a valid pointer (since it can't find the "real" start of the allocation) and then it will fail in a way that'd make it easier to detect.

**Comment by basvodde** [comment]
Yeah, but the separate tracking will actually cause performance issues as it doubles the number of allocations. I guess it could be changed in a compiler flag.

It would be useful to figure out what call is allocating and deallocating in library. But it is hard to do anything about it :(

**Comment by uecasm** [comment]
It doesn't need to double the number of allocations.

Just allocate one large block, as now, but instead of using the end of it for the tracking data and handing out the original base pointer to the user code, do it the other way around.  You have to be a little more careful about alignment, but that's fairly minor.

The issue is that if someone takes a pointer allocated by CppUTest and then passes it to the regular delete instead of CppUTest's override, then they're still passing it to the "right" underlying allocator (since that lives in the C++ runtime shared by everything), and it sees a valid pointer to the start of the memory block, so it happily deallocates it.  Meanwhile it's still in CppUTest's linked list of memory.  Oops.

If you pass a pointer to somewhere in the middle of the "real" allocated block (which is the only pointer the user code would have, since it doesn't know how to find the real block start without CppUTest's help), then the allocator will usually signal an error and refuse to deallocate it (because it sees it as not a valid block).  Which is also not great, but it's probably less harmful, and will still end up getting flagged as a memory leak by CppUTest (if not caught earlier by some other allocation checker, eg. in debug mode).

**Comment by basvodde** [comment]
Yeah, true. That would work. I'll keep this open as that might be worth the effort.
